pims/base_frames.py

```
@@ -528,3 +528,267 @@ def process(img_or_iterable, *args, **kwargs):
+        ndarray with shape `frame_shape_2D` and dtype `pixel_type`.
+        """
+        pass
+
+    def get_frame(self, i):
+        """ Returns a Frame of shape deterimend by aggregate. The property
+        iterate (together with the dimension default index) determines which.
+        multidimensional index is returned. """
+        # identify the indices to take along each dimension
+        i_prev = 0
+        for n, dim in enumerate(self._dims):
+            dim.i = dim.default
+            if dim.iterate:
+                size = int(np.prod([d.size for d in self._dims[n+1:]
+                                    if d.iterate]))
+                dim.i = (i - i_prev) // size
+                i_prev += dim.i * size
+
+        # initialize a stack of 2D arrays to collect the Frame
+        Nframes = int(np.prod(self.frame_shape[:-2]))
+        result = np.empty([Nframes] + list(self.frame_shape_2D),
+                          dtype=self.pixel_type)
+
+        # read all 2D frames
+        for n in range(Nframes):
+            result[n] = self.get_frame_2D(**{d.name: d.i for d in self._dims})
+            for dim in self._dims[::-1]:
+                if dim.aggregate:
+                    dim.i += 1
+                    if dim.i >= dim.size:
+                        dim.i = 0
+                    else:
+                        break
+
+        # reshape the array into the desired shape
+        result.shape = self.frame_shape
+        return Frame(result, frame_no=i)
+
+    def __getattr__(self, key):
+        """ Sets an empty value for self._dims without using __init__.
+
+        Enables dimensions to be accessed by for instance frames.c.
+        Existing methods always precede over this: a dimension named 'ndim'
+        would not be accessible by frames.ndim. """
+        if key == '_dims':
+            return []
+        for dim in self._dims:
+            if key == dim.name:
+                return dim
```

I think this approach fails silently if I try to access a dim that does not exist.